From 4bf62b339ef3a30dac10160da76905222063485b Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Fri, 27 Jun 2008 02:48:06 +0000
Subject: [PATCH] Narrowed synchronization in AbandonedTrace to resolve an
 Evictor deadlock. JIRA: DBCP-270 Reported and patched by Filip Hanik.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/trunk@672097 13f79535-47bb-0310-9956-ffa450edef68
---
 .../org/apache/commons/dbcp/AbandonedTrace.java  | 16 +++++++++-------
 xdocs/changes.xml                                |  3 +++
 2 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/src/java/org/apache/commons/dbcp/AbandonedTrace.java b/src/java/org/apache/commons/dbcp/AbandonedTrace.java
index efcef69fed..b33b0494fd 100644
--- a/src/java/org/apache/commons/dbcp/AbandonedTrace.java
+++ b/src/java/org/apache/commons/dbcp/AbandonedTrace.java
@@ -172,7 +172,7 @@ protected void setStackTrace() {
      * @param trace AbandonedTrace object to add
      */
     protected void addTrace(AbandonedTrace trace) {
-        synchronized (this) {
+        synchronized (this.trace) {
             this.trace.add(trace);
         }
         setLastUsed();
@@ -182,8 +182,8 @@ protected void addTrace(AbandonedTrace trace) {
      * Clear the list of objects being traced by this
      * object.
      */
-    protected synchronized void clearTrace() {
-        if (this.trace != null) {
+    protected void clearTrace() {
+        synchronized(this.trace) {
             this.trace.clear();
         }
     }
@@ -194,7 +194,9 @@ protected synchronized void clearTrace() {
      * @return List of objects
      */
     protected List getTrace() {
-        return trace;
+        synchronized (this.trace) {
+            return new ArrayList(trace);
+    }
     }
 
     /**
@@ -206,7 +208,7 @@ public void printStackTrace() {
             System.out.println(format.format(new Date(createdTime)));
             createdBy.printStackTrace(System.out);
         }
-        synchronized(this) {
+        synchronized(this.trace) {
             Iterator it = this.trace.iterator();
             while (it.hasNext()) {
                 AbandonedTrace at = (AbandonedTrace)it.next();
@@ -220,8 +222,8 @@ public void printStackTrace() {
      *
      * @param trace AbandonedTrace object to remvoe
      */
-    protected synchronized void removeTrace(AbandonedTrace trace) {
-        if (this.trace != null) {
+    protected void removeTrace(AbandonedTrace trace) {
+        synchronized(this.trace) {
             this.trace.remove(trace);
         }
     }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index b937e40c0e..e658bc268b 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -42,6 +42,9 @@ The <action> type attribute can be add,update,fix,remove.
      new features as well as bug fixes and instrumentation.  Some bug fixes
      will change semantics (e.g. connection close will become idempotent).
      The minimum JDK level will be increased to 1.4">
+      <action dev="psteitz" type="fix" issue="DBCP-270" due-to="Filip Hanik">
+        Narrowed synchronization in AbandonedTrace to resolve an Evictor deadlock.
+      </action>
       <action dev="bayard" type="fix" issue="DBCP-218">
         Corrected Javadoc to state that getLoginTimeout and setLoginTimeout are
         NOT supported by BasicDataSource.
